### PR TITLE
Add NNC AOT Compiler executable

### PR DESCRIPTION
**Pull request by priyaramani**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63994

Differential Revision: [D30582149](https://our.internmc.facebook.com/intern/diff/D30582149)

Goal: Add ability to do Ahead Of Time (AOT) compilation of models. 
This PR adds an API to compile models AOT . `aot_model_compiler` cpp executable will compile the forward method of the given model and produce LLVM assembly code and compiled model files as output.

On a bigger picture, the goal of AOT compilation is to be able to achieve extremely small pytorch binary size. 
As a bonus, we get better performance with a faster runtime too.

Thanks Mikhail and Jiakai for the prototype! 

Steps build aot_compiler and compile mobilenetv3 model:
```
# Build PyTorch with LLVM
USE_LLVM=<LLVM install path> python setup.py develop

# Get mobilenetv3 model
jf download GICWmACPJxxBk_YQAO8qqJoT-VkMbsIXAAAz --file mobilenetv3.pt

# Build aot_model_compiler cpp executable 
./scripts/build_local.sh -DBUILD_BINARY=ON

# Run aot_model_compiler on mobilenetv3 model
build/bin/aot_model_compiler --model mobilenetv3.pt --model_name=pytorch_dev_mobilenetv3 --model_version=v1 --input_dims="1,3,224,224"
```

Output:
```
Input sizes: 1, 3, 224, 224,
graph(%x.1 : Tensor):
  %2 : Float(1280, 1000, strides=[1, 1280], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %3 : int[] = prim::Constant[value=[2]]()
  %4 : int[] = prim::Constant[value=[3]]()
  %5 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %6 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %7 : Float(144, 576, strides=[1, 144], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %8 : Float(576, 144, strides=[1, 576], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %9 : Float(72, 288, strides=[1, 72], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
  %10 : Float(288, 72, strides=[1, 288], requires_grad=0, device=cpu) = prim::Constant[value=<Tensor>]()
.
.
.
.
 %self.features.14.bias, %24, %23, %24, %29) # /home/mvz/pytorch/torch/nn/modules/conv.py:386:15
  %420 : Tensor(1, 1280, 1, 1) = aten::add(%input14.1, %26, %29) # mobilenetv3.py:31:27
  %421 : Tensor(1, 1280, 1, 1) = aten::hardtanh(%420, %28, %27) # /home/mvz/pytorch/torch/nn/functional.py:1161:17
  %422 : Tensor(1, 1280, 1, 1) = aten::mul(%input14.1, %421) # mobilenetv3.py:31:15
  %x0.1 : Tensor(1, 1280, 1, 1) = aten::div(%422, %27) # mobilenetv3.py:31:15
  %424 : Tensor(1, 1280, 1) = aten::mean(%x0.1, %4, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %x1.1 : Tensor(1, 1280) = aten::mean(%424, %3, %self.features.0.1.training, %self.features.0.0.bias) # mobilenetv3.py:205:12
  %ret0.1 : Tensor(1, 1000) = aten::addmm(%self.classifier.1.bias, %x1.1, %2, %29, %29) # /home/mvz/pytorch/torch/nn/functional.py:1665:14
  return (%ret0.1)

The compiled model was saved to mobilenetv3.compiled.pt
```

Outputs compiled model and LLVM assembly code files
```
(pytorch)  ~/local/pytorch aot
└─ $ ls | grep mobilenetv3.compiled
mobilenetv3.compiled.ll
mobilenetv3.compiled.pt
```

Further integration tests at [D30595123](https://our.internmc.facebook.com/intern/diff/D30595123)

### CODE REVIEW

File: torch/csrc/jit/mobile/nnc/aot_compiler.cpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    out_spec.push_back(output);
+  }
+  func->set_output_specs(out_spec);
+}
+
+std::unique_ptr<Function> aot_compile(
+    const std::string& method_name,
+    std::shared_ptr<Graph>& g,
+    const std::vector<int64_t>& sizes,
+    std::string* compiled_assembly) {
```
Comment (bertmaher) [inline torch/csrc/jit/mobile/nnc/aot_compiler.cpp:78]: Minor: slight preference for returning a `pair` or a custom struct rather than a pointer out param.

**Review by bertmaher — APPROVED** [review]
Looks good, only minor comments from me.

File: torch/csrc/jit/mobile/nnc/aot_compiler.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
+// (c) Facebook, Inc. and its affiliates. Confidential and proprietary.
```
Comment (ZolotukhinM) [inline torch/csrc/jit/mobile/nnc/aot_compiler.cpp:1]: Is it confidential and proprietary though?

File: torch/csrc/jit/mobile/nnc/aot_compiler.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+#include <ATen/NativeFunctions.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_opt_limit.h>
+#include <torch/csrc/jit/passes/constant_propagation.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+#include <torch/csrc/jit/tensorexpr/ir.h>
+#include <torch/csrc/jit/tensorexpr/ir_simplifier.h>
+#include <torch/csrc/jit/tensorexpr/kernel.h>
+#include <torch/csrc/jit/tensorexpr/llvm_codegen.h>
+#include <torch/csrc/jit/tensorexpr/loopnest.h>
```
Comment (ZolotukhinM) [inline torch/csrc/jit/mobile/nnc/aot_compiler.cpp:15]: Some includes here are unnecessary:
```
ir_simplifier.h
loopnest.h
llvm_codegen.h
jit_opt_limit.h
```
It's a good practice only include what's used in the file. Also, please run clang-format to sort includes alphabetically.

File: torch/csrc/jit/mobile/nnc/aot_compiler.cpp [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    std::cerr << s << ", ";
+  }
+  std::cerr << "\n";
+
+  RemoveTensorMutation(g);
+  EliminateDeadCode(g->block());
+  g = tensorexpr::removeUnusedSelfArgument(g);
+
+  std::vector<c10::optional<at::Tensor>> example_inputs = {at::rand(sizes)};
+  g->dump();
```
Comment (ZolotukhinM) [inline torch/csrc/jit/mobile/nnc/aot_compiler.cpp:91]: We probably don't need to dump the graph here. If you do want to keep debugging statements, please use `GRAPH_DUMP(g)` - this macro can be toggled on/off with `PYTORCH_JIT_LOG_LEVEL` environment variable (see more details on this here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/OVERVIEW.md#jit-logging).

File: torch/csrc/jit/mobile/nnc/aot_compiler.cpp [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    const std::string& method_name,
+    std::shared_ptr<Graph>& g,
+    const std::vector<int64_t>& sizes,
+    std::string* compiled_assembly) {
+  auto g2 = g->copy();
+  std::cerr << "Input sizes: ";
+  for (auto s : sizes) {
+    std::cerr << s << ", ";
+  }
+  std::cerr << "\n";
```
Comment (ZolotukhinM) [inline torch/csrc/jit/mobile/nnc/aot_compiler.cpp:84]: Let's cleanup debug prints, or wrap them with `GRAPH_DEBUG` macros.

File: torch/csrc/jit/mobile/nnc/aot_compiler.cpp [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+  std::vector<c10::optional<at::Tensor>> example_inputs = {at::rand(sizes)};
+  g->dump();
+  tensorexpr::annotateInputShapes(g, example_inputs);
+
+  PropagateShapesOnGraph(g);
+  PeepholeOptimize(g, false);
+  ConstantPropagation(g);
+  PropagateShapesOnGraph(g);
+  g->dump();
```
Comment (ZolotukhinM) [inline torch/csrc/jit/mobile/nnc/aot_compiler.cpp:98]: Same

File: torch/csrc/jit/mobile/nnc/aot_compiler.h [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+#include <torch/csrc/jit/mobile/nnc/context.h>
+#include <torch/csrc/jit/ir/ir.h>
+
+namespace torch {
+namespace jit {
+namespace mobile {
+namespace nnc {
+
+
+TORCH_API std::unique_ptr<Function> aot_compile(
```
Comment (ZolotukhinM) [inline torch/csrc/jit/mobile/nnc/aot_compiler.h:14]: Please add a brief comment for the function.

File: torch/csrc/jit/mobile/nnc/aot_compiler.cpp [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+  for (auto dim : b->dims()) {
+    IntImmPtr int_imm_dim = to<IntImm>(dim);
+    // TODO: assert it's actually immediate
+    int64_t s = int_imm_dim->value();
+    r.push_back(s);
+  }
+  return r;
+}
+
+void get_compiled_function(
```
Comment (ZolotukhinM) [inline torch/csrc/jit/mobile/nnc/aot_compiler.cpp:41]: Nit: even though we're not very consistent with this, I think our guidelines are to use lowerCamelCase for function names:
> * We follow the Facebook C++ Style Guidelines. Examples: some_namespace, SomeClass, SomeEnum, someFunction, someMethod, someMemberVariable, somePrivateMemberVariable_, SOME_MACRO
> * Exception: C++ methods and functions bound into python use python naming scheme (snake_case, ...) also in their C++ names. Exposing a previously unexposed method might need renaming the C++ method.

**Review by ZolotukhinM — APPROVED** [review]
👍

**Comment by priyaramani** [comment]
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

File: torch/csrc/jit/mobile/nnc/aot_compiler.cpp (hunk 8 above)
Comment (priyaramani) [reply torch/csrc/jit/mobile/nnc/aot_compiler.cpp:41]: Thanks for the review! 
Addressed the comments.

File: binaries/aot_model_compiler.cc [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+C10_DEFINE_string(
+    input_dims,
+    "",
+    "Alternate to input_files, if all inputs are simple "
+    "float TensorCPUs, specify the dimension using comma "
+    "separated numbers. If multiple input needed, use "
+    "semicolon to separate the dimension of different "
+    "tensors.");
+C10_DEFINE_string(
+    output_asm,
```
[the remark this answers is not in the document]
Comment (priyaramani) [reply binaries/aot_model_compiler.cc:27]: Thanks for the review!
Addressed the comments.

**Comment by bertmaher** [comment]
@IvanKobzarev it looks like somehow reverted this PR to an earlier version?

**Comment by IvanKobzarev** [comment]
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

**Comment by IvanKobzarev** [comment]
@ivankobzarev has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

**Comment by priyaramani** [comment]
> @IvanKobzarev it looks like somehow reverted this PR to an earlier version?

**Comment by priyaramani** [comment]
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

**Comment by priyaramani** [comment]
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

**Comment by priyaramani** [comment]
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

**Comment by priyaramani** [comment]
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

**Comment by priyaramani** [comment]
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

**Comment by priyaramani** [comment]
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30582149).

**Pull request closed without merging**